### PR TITLE
fix detach azure disk back off issue which has too big lock in failure retry condition

### LINKED ISSUE #76502: detach azure disk should not hold data disk list for a long time
**What happened**:
In some error condition when detach azure disk failed, azure cloud provider will retry 6 times at most with exponential backoff, it will hold the data disk list for about 3 minutes with a node level lock, and in that time period, if customer update data disk list manually (e.g. need manual operationto attach/detach another disk since there is attach/detach error, ) , the data disk list will be obselete(dirty data), then weird VM status happens, e.g. attach a non-existing disk, we should split those retry operations, every retry should get a fresh data disk list in the beginning.

https://github.com/kubernetes/kubernetes/blob/cbaaa67b11ee49fcc1ca1dcb33a7e9e1025cee81/pkg/cloudprovider/providers/azure/azure_controller_standard.go#L150-L153

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): master
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig azure
/assign

### PR DESCRIPTION
**Pull request by andyzhangx**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
In some error condition when detach azure disk failed, azure cloud provider will retry 6 times at most with exponential backoff, it will hold the data disk list for about 3 minutes with a node level lock, and in that time period, if customer update data disk list manually (e.g. need manual operationto attach/detach another disk since there is attach/detach error, ) , the data disk list will be obselete(dirty data), then weird VM status happens, e.g. attach a non-existing disk, we should split those retry operations, every retry should get a fresh data disk list in the beginning.

https://github.com/kubernetes/kubernetes/blob/cbaaa67b11ee49fcc1ca1dcb33a7e9e1025cee81/pkg/cloudprovider/providers/azure/azure_controller_standard.go#L150-L153

This PR has two commits:
 - [refactor detach azure disk retry operation](https://github.com/kubernetes/kubernetes/commit/39c239c308b6ae79e014c755df9d1ddfbc3a0499): 
1. rename function name from `DetachDiskByName` to `DetachDisk`
2. refine detach azure disk retry operation, make every detach azure disk operation in a standalone function, originally it's by `as.CreateOrUpdateVMWithRetry(nodeResourceGroup, vmName, newVM)` which may lead to obsolete data disk list

 - [move disk lock process to azure cloud provider](https://github.com/kubernetes/kubernetes/commit/299ef429830caf347814951d9957a6340ae70b74):
Thus there would be separate lock for every detach azure disk operation, we don't need to lock whole disk detach backoff(6 retries) process.

This PR don't change the logic of attach disk since there is no retry in azure cloud provide for attach disk, k8s attach-detach controller will do the attach volume retry.

BTW, I have tested this PR on both vmss and vmas k8s cluster by a stress disk attach/detach test, it works well for a bunch of times.

**Which issue(s) this PR fixes**:

Fixes #76502

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
fix detach azure disk back off issue which has too big lock in failure retry condition
```

/kind bug
/assign @feiskyer 
/priority important-soon
/sig azure

cc @khenidak @brendandburns

### CODE REVIEW

File: pkg/cloudprovider/providers/azure/azure_controller_common.go [hunk 1]
```diff
@@ -85,24 +90,72 @@ func (c *controllerCommon) getNodeVMSet(nodeName types.NodeName) (VMSet, error)
...
+	}
+
+	klog.V(2).Infof("Trying to attach volume %q lun %d to node %q.", diskURI, lun, nodeName)
 	return vmset.AttachDisk(isManagedDisk, diskName, diskURI, nodeName, lun, cachingMode)
 }
 
-// DetachDiskByName detaches a vhd from host. The vhd can be identified by diskName or diskURI.
-func (c *controllerCommon) DetachDiskByName(diskName, diskURI string, nodeName types.NodeName) error {
+// DetachDisk detaches a disk from host. The vhd can be identified by diskName or diskURI.
+func (c *controllerCommon) DetachDisk(diskName, diskURI string, nodeName types.NodeName) error {
```
Comment (feiskyer) [inline pkg/cloudprovider/providers/azure/azure_controller_common.go]: Add some unit tests for DetachDisk and AttachDisk?

Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_controller_common.go]: done, it's under `azure_controller_common_test.go`

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Comment by feiskyer** [comment]
@andrewsykim Could you help to approve the cloud-provider changes?

/assign @andrewsykim

**Comment by andrewsykim** [comment]
Can we add a bit more details on the release note please?

**Comment by andyzhangx** [comment]
@andrewsykim thanks. I have changed the release note to:
```release-note
fix detach azure disk back off issue which has too big lock in failure retry condition
```
Let me know if you have any question.

**Comment by andyzhangx** [comment]
@andrewsykim PTAL, thanks.

**Comment by andrewsykim** [comment]
/approve
/lgtm 

for  `pkg/cloudprovider/providers/.import-restrictions` changes

**Pull request merged**